src/ReinforcementLearningZoo/src/algorithms/policy_gradient/vpg.jl

```diff
@@ -1,147 +1,87 @@
-)
-    model = π.approximator
-    to_dev(x) = send_to_device(device(model), x)
+function RLBase.optimise!(p::VPG, batch::NamedTuple{(:state, :action, :gain)})
+    A = p.approximator
+    B = p.baseline
+    s, a, g = map(Array, batch) # !!! FIXME
 
-    states = traj[:state]
-    actions = traj[:action] |> Array # need to convert ElasticArray to Array, or code will fail on gpu. `log_prob[CartesianIndex.(A, 1:length(A))`
-    gains = traj[:reward] |> x -> discount_rewards(x, π.γ)
-
-    for idx in Iterators.partition(shuffle(1:length(traj[:terminal])), π.batch_size)
-        S = select_last_dim(states, idx) |> Array |> to_dev
-        A = actions[idx]
-        G = gains[idx] |> x -> Flux.unsqueeze(x, dims=1) |> to_dev
-        # gains is a 1 column array, but the output of flux model is 1 row, n_batch columns array. so unsqueeze it.
-
-        if π.baseline isa NeuralNetworkApproximator
-            gs = gradient(Flux.params(π.baseline)) do
-                δ = G - π.baseline(S)
-                loss = mean(δ .^ 2) * π.α_w # mse
-                ignore_derivatives() do
-                    π.baseline_loss = loss
-                end
-                loss
-            end
-            update!(π.baseline, gs)
-        elseif π.baseline isa Nothing
-            # Normalization. See
-            # (http://rail.eecs.berkeley.edu/deeprlcourse-fa17/f17docs/hw2_final.pdf)
-            # (https://web.stanford.edu/class/cs234/assignment3/solution.pdf)
-            # normalise should not be used with baseline. or the loss of the policy will be too small.
-            δ = G |> x -> normalise(x; dims=2)
-        end
-
-        gs = gradient(Flux.params(model)) do
-            if π.action_space isa AbstractVector
-                log_prob = S |> model |> logsoftmax
-                log_probₐ = log_prob[CartesianIndex.(A, 1:length(A))]
-            elseif π.action_space isa Interval
-                dist = π.dist.(model(S)...) # TODO: this part does not work on GPU. See: https://github.com/JuliaStats/Distributions.jl/issues/1183 .
-                log_probₐ = logpdf.(dist, A)
-            end
-            loss = -mean(log_probₐ .* δ) * π.α_θ
+    if isnothing(B)
+        δ = normalise(g)
+    else
+        gs = gradient(params(B)) do
+            δ = g - vec(B(s))
+            loss = mean(δ .^ 2)
             ignore_derivatives() do
-                π.loss = loss
+                # @info "VPG/baseline" loss = loss δ
             end
             loss
         end
-        update!(model, gs)
+        optimise!(B, gs)
+    end
+
+    gs = gradient(params(A)) do
+        if p.dist <: DiscreteDistribution
+            log_prob = s |> A |> logsoftmax
+            log_probₐ = log_prob[CartesianIndex.(a, 1:length(a))]
+        elseif p.dist <: ContinuousDistribution
+            dist = p.dist.(A(s)...) # TODO: this part does not work on GPU. See: https://github.com/JuliaStats/Distributions.jl/issues/1183 .
+            log_probₐ = logpdf.(dist, A)
+        end
```

Comment:
i'm not sure this would work.  `dist` maps the output of the network to a distribution, so it's a function type in general.  in the discrete case, where `dist = Categorical`, `Categorical` is coincidentally the type as well as the constructor, so `p.dist <: DiscreteDistribution` would match.  but in general, `dist` need not be a `Distribution` type.

Reply:
Yes, here `dist` is a mapping from `parameters... -> distribution`. A more conventional way is to add a custom trait to specify the distribution type of the result. But given that most distributions provided in `Distributions.jl` are both type and function, I think it's safe to assume `dist` is always a struct type here.

Reply:
oh!  the example i had in mind was `Normal(μ,σ)` which has two arguments but i guess that works too with `A(s)...`, nice